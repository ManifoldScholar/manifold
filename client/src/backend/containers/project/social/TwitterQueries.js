import React, { Component } from "react";
import PropTypes from "prop-types";
import { withTranslation } from "react-i18next";
import { entityStoreActions } from "actions";
import { projectsAPI, requests } from "api";
import { select, meta } from "utils/entityUtils";
import lh from "helpers/linkHandler";
import { connect } from "react-redux";
import get from "lodash/get";
import EntitiesList, {
  Button,
  TwitterQueryRow
} from "backend/components/list/EntitiesList";

import Authorize from "hoc/Authorize";

const { request } = entityStoreActions;
const perPage = 10;

export class ProjectSocialTwitterQueriesContainer extends Component {
  static mapStateToProps = state => {
    return {
      twitterQueries: select(requests.beTwitterQueries, state.entityStore),
      twitterQueriesMeta: meta(requests.beTwitterQueries, state.entityStore)
    };
  };

  static displayName = "Project.Social.TwitterQueries";

  static propTypes = {
    project: PropTypes.object,
    twitterQueries: PropTypes.array,
    twitterQueriesMeta: PropTypes.object,
    match: PropTypes.object,
    dispatch: PropTypes.func,
    t: PropTypes.func
  };

  componentDidMount() {
    this.fetchTwitterQueries(1);
  }

  fetchTwitterQueries(page) {
    const pagination = { number: page, size: perPage };
    const action = request(
      projectsAPI.twitterQueries(this.props.project.id, pagination),
      requests.beTwitterQueries
    );
    this.props.dispatch(action);
  }

  handlePageChange(event, page) {
    this.fetchTwitterQueries(page);
  }

  pageChangeHandlerCreator = page => {
    return event => {
      this.handlePageChange(event, page);
    };
  };

  render() {
    const { twitterQueries, twitterQueriesMeta, project, t } = this.props;
    if (!twitterQueriesMeta || !twitterQueries) return null;

    const active = get(this.props.match, "params.qId");

    return (
      <Authorize entity={project} ability="manageTwitterQueries">
        <EntitiesList
          entityComponent={TwitterQueryRow}
          entities={twitterQueries}
          entityComponentProps={{ active }}
          title={t("projects.twitter_queries")}
          titleStyle="section"
          instructions={t("projects.twitter_queries_instructions")}
          buttons={[
            <Button
              path={lh.link("backendProjectSocialTwitterQueryNew", project.id)}
              text={t("projects.add_twitter_label")}
              type="add"
              authorizedFor={project}
              authorizedTo="createTwitterQueries"
            />
          ]}
        />
      </Authorize>
    );
  }
}

export default withTranslation()(
  connect(ProjectSocialTwitterQueriesContainer.mapStateToProps)(
    ProjectSocialTwitterQueriesContainer
  )
);
