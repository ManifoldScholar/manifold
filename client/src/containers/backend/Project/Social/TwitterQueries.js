import React, { Component } from "react";
import PropTypes from "prop-types";
import { List, TwitterQuery, Form } from "components/backend";
import { entityStoreActions } from "actions";
import { projectsAPI, requests } from "api";
import { select, meta } from "utils/entityUtils";
import { Link } from "react-router-dom";
import { Utility } from "components/global";
import lh from "helpers/linkHandler";
import { connect } from "react-redux";
import get from "lodash/get";

const { request } = entityStoreActions;
const perPage = 10;

export class ProjectSocialTwitterQueriesContainer extends Component {
  static displayName = "Project.Social.TwitterQueries";

  static mapStateToProps = state => {
    return {
      twitterQueries: select(requests.beTwitterQueries, state.entityStore),
      twitterQueriesMeta: meta(requests.beTwitterQueries, state.entityStore)
    };
  };

  static propTypes = {
    project: PropTypes.object,
    twitterQueries: PropTypes.array,
    twitterQueriesMeta: PropTypes.object,
    match: PropTypes.object,
    dispatch: PropTypes.func
  };

  componentDidMount() {
    this.fetchTwitterQueries(1);
  }

  fetchTwitterQueries(page) {
    this.lastFetchedPage = page;
    const pagination = { number: page, size: perPage };
    const action = request(
      projectsAPI.twitterQueries(this.props.project.id, pagination),
      requests.beTwitterQueries
    );
    this.props.dispatch(action);
  }

  handlePageChange(event, page) {
    this.fetchTwitterQueries(page);
  }

  pageChangeHandlerCreator = page => {
    return event => {
      this.handlePageChange(event, page);
    };
  };

  render() {
    const { twitterQueries, twitterQueriesMeta } = this.props;
    if (!twitterQueriesMeta || !twitterQueries) return null;

    const active = get(this.props.match, "params.qId");
    const instructions = (
      <p className="instructional-copy">
        Manifold will periodically fetch tweets according to the queries
        specified below.
      </p>
    );

    return (
      <Form.FieldGroup label="Twitter Queries" instructions={instructions}>
        <nav className="vertical-list-primary flush">
          <div className="buttons-icon-horizontal">
            <Link
              to={lh.link(
                "backendProjectSocialTwitterQueryNew",
                this.props.project.id
              )}
              className="button-icon-secondary"
            >
              <i className="manicon manicon-plus" />
              Add a New Twitter Query
            </Link>
          </div>
          {twitterQueries
            ? <List.SimpleList
                entities={twitterQueries}
                entityComponent={TwitterQuery.ListItem}
                entityComponentProps={{ active }}
              />
            : null}
        </nav>
        <Utility.Pagination
          pagination={twitterQueriesMeta.pagination}
          paginationClickHandler={this.pageChangeHandlerCreator}
        />
      </Form.FieldGroup>
    );
  }
}

export default connect(ProjectSocialTwitterQueriesContainer.mapStateToProps)(
  ProjectSocialTwitterQueriesContainer
);
