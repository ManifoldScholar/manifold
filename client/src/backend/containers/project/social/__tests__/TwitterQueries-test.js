import { ProjectSocialTwitterQueriesContainer } from "../TwitterQueries";

describe("backend/containers/project/social/TwitterQueries", () => {
  def("pagination", () => fixtures.pagination());
  def("twitterQueries", () => collectionFactory("twitterQuery"));
  def("project", () => factory("project"));
  def("match", () => ({ params: { qId: "twitterQuery-2" } }));

  def("root", () => (
    <ProjectSocialTwitterQueriesContainer
      project={$project}
      dispatch={$dispatch}
      twitterQueries={$twitterQueries}
      twitterQueriesMeta={{
        pagination: $pagination
      }}
      match={$match}
      t={key => key}
    />
  ));

  it("matches the snapshot when rendered", () => {
    expect(render($withApp($root)).html()).toMatchSnapshot();
  });
});
