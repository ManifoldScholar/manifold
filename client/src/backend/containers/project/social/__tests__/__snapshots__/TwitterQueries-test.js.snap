// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend Project Social Twitter Queries Container renders correctly 1`] = `
<div
  className="entity-list"
  id="entities-list1"
>
  <header
    className="entity-list__title-block entity-list__title backend-header"
  >
    <h2
      className="backend-header__h2-title"
    >
      Twitter Queries
    </h2>
  </header>
  <div
    className="entity-list__contents-wrapper entity-list__contents-wrapper--with-section-title"
  >
    <div
      className="instructional-copy"
    >
      
            Manifold will periodically fetch tweets according to the queries specified
            below.
          
    </div>
    <div
      className="entity-list__header"
    >
      <div
        className="entity-list__button-set"
      >
        <a
          className="entity-list__button"
          href="/backend/projects/1/social/twitter-query/new"
          onClick={[Function]}
        >
          <svg
            aria-hidden={true}
            className="manicon-svg svg-icon--circlePlus32"
            fill="currentColor"
            height={null}
            viewBox="0 0 32 32"
            width={null}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.99999998,16.5 L9.99999998,15.5 L22,15.5 L22,16.5 L9.99999998,16.5 Z M16.5,22 L15.5,22 L15.5,9.99999998 L16.5,9.99999998 L16.5,22 Z M16,27 C9.92486773,27 4.99999997,22.0751323 4.99999997,16 C4.99999997,9.92486773 9.92486773,4.99999997 16,4.99999997 C22.0751323,4.99999997 27,9.92486773 27,16 C27,22.0751323 22.0751323,27 16,27 Z M16,26 C21.5228475,26 26,21.5228475 26,16 C26,10.4771525 21.5228475,5.99999997 16,5.99999997 C10.4771525,5.99999997 5.99999997,10.4771525 5.99999997,16 C5.99999997,21.5228475 10.4771525,26 16,26 Z"
            />
          </svg>
          Add a new twitter query
        </a>
      </div>
    </div>
    <ul
      className="entity-list__list entity-list__list--rows"
    >
      <li
        className="entity-row entity-list__entity"
      >
        <a
          aria-describedby="undefined1-describedby"
          className="entity-row__row-link entity-row__row-link--block entity-row__row-link--atag entity-row__row-link--is-active"
          href="/backend/projects/1/social/twitter-query/2"
          onClick={[Function]}
        >
          <div
            className="entity-row__inner entity-row__inner--in-rows entity-row__inner--with-row-link"
          >
            <div
              className="entity-row__text entity-row__text--in-rows"
            >
              <h4
                className="entity-row__title entity-row__title--in-rows"
              >
                <span
                  className="entity-row__title-inner"
                >
                  from:manifoldscholar
                </span>
                <span
                  className="entity-row__labels"
                >
                  <span
                    className="entity-row__label"
                  >
                    Active
                  </span>
                </span>
                <span
                  className="aria-describedby"
                  id="undefined1-describedby"
                >
                  View from:manifoldscholar
                </span>
              </h4>
              <h6
                className="entity-row__meta entity-row__meta--in-rows"
              >
                No tweets have been fetched yet.
              </h6>
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
